src/modules/adaptors/Balancer/BalancerPoolAdaptor.sol

```diff
@@ -0,0 +1,295 @@
+     * @notice This position is a liquidity provision (credit) position, and user withdraws are not allowed so this position must return 0 for withdrawableFrom.
+     */
+    function withdrawableFrom(bytes memory, bytes memory) public pure override returns (uint256) {
+        return 0;
+    }
+
+    /**
+     * @notice Returns the cellars balance of the positions debtToken.
+     * TODO: make sure it is checking balanceOf && staked tokens
+     */
+    function balanceOf(bytes memory adaptorData) public view override returns (uint256) {
+        // TODO: decode adaptorData for poolGauge address
+        address bpt = abi.decode(adaptorData, (address));
+
+        // ILiquidityGaugev3Custom poolGauge = ILiquidityGaugev3Custom(_poolGauge); // TODO: usually poolGauge address is provided, but in this case we'll have to decode.
+
+        uint256 stakedBPT = poolGauge.balanceOf(address(this));
+        return ERC20(token).balanceOf(msg.sender) + stakedBPT;
+    }
+
+    /**
+     * @notice Returns the positions debtToken underlying asset.
+     */
+    function assetOf(bytes memory adaptorData) public view override returns (ERC20) {
+        IERC20 token = IERC20(abi.decode(adaptorData, (address))); // TODO: sort out this decoding and the adaptorData contents
+
+        return ERC20(token); // TODO: confirm if this should be all underlying assets or just the bpt.
+    }
+
+    /**
+     * @notice This adaptor returns collateral, and not debt.
+     * @return whether adaptor returns debt or not
+     */
+    function isDebt() public pure override returns (bool) {
+        return false;
+    }
+
+    //============================================ Strategist Functions ===========================================
+
+    /**
+     * @notice call `BalancerRelayer` on mainnet to carry out txs that can incl. joins, exits, and swaps
+     * @param tokensIn specific tokens being input for tx
+     * @param amountsIn amount of assets input for tx
+     * @param bptOut acceptable amount of assets resulting from tx (due to slippage, etc.)
+     * @param callData encoded specific txs to be used in `relayer.multicall()`
+     * NOTE: The difference btw what Balancer outlines to do and what we want to do is that we want the txs to be carried out by the vault. Whereas we could just end up connecting to the relayer and have it carry out multi-txs which would be bad cause we need to confirm each step of the multi-call.
+     * TODO: rename params and possibly include others for `exit()` function. ex.) bptOut needs to be renamed to be agnostic.
+     * TODO: see issue for strategist callData
+     */
+    function useRelayer(
+        ERC20[] memory tokensIn,
+        uint256[] memory amountsIn,
+        ERC20 bptOut,
+        bytes[] memory callData
+    ) public {
+        for (uint256 i; i < tokensIn.length; ++i) tokensIn[i].approve(relayer(), amountsIn[i]); // TODO: this may be obsolete based on approval steps necessary for relayer functionality (see TODO below)
```

Comment:
Hmm, I think we only need to approve the relayer because it would take our tokens and act on our behalf on the vault right? 

Referencing the docs, it states the relayer is the single point of entry [here](https://docs.balancer.fi/concepts/advanced/relayers.html#how-it-works:~:text=which%20is%20the%20single%20point%20of%20entry%20via%20the%20multicall%20function), and that a user has to apprve each relayer before use [here](https://docs.balancer.fi/concepts/advanced/relayers.html#contracts:~:text=A%20user%20has%20to%20approve%20each%20Relayer%20before%20they%20can%20use%20it.%20To%20check%20if%20a%20Relayer%20is%20approved%20the%20hasApprovedRelayer%20on%20the%20Vault%20can%20be%20used.)

From a quick skim on the code, it looks like the `Relayer` calls `multiCall` which acts upon the `Vault` where things like `joinPool()` could be called. Going down that path more, `joinPool()` shows that transfers eventually happen from `AssetTransfersHandler.sol` (Vault inherits this contract). It is a `safeTransferFrom()` / typical transfer calls. So the `sender` mentioned in this call, is I think the original caller - so the `Cellar` in our case. If we call `relayerApproval` then we are giving `allowance` to the relayer to spend on the Cellar behalf. 

So here's a question with all that: is `msg.sender` the `Cellar` address if we call it using `delegateCall()` from the `BalancerPoolAdaptor` in this eventual function call in `BaseRelayerLibrary.sol`? I think that it is. So we'd be giving approval to the relayer to spend the `Cellar` assets. At the end we'd have to revoke approval for the Relayer on behalf of the Cellar, so change this [value](https://github.com/balancer/balancer-v2-monorepo/blob/21e028ca8fe31beff6c90e007739b85ce8d8c4c6/pkg/vault/contracts/VaultAuthorization.sol#LL60C5-L60C76). I'm not sure how to do this because it doesn't show a way to do so in the aforementioned Balancer contracts. 

```
function setRelayerApproval(
        address relayer,
        bool approved,
        bytes calldata authorisation
    ) external payable {
        require(relayer == address(this) || !approved, "Relayer can only approve itself");
        bytes memory data = abi.encodePacked(
            abi.encodeWithSelector(_vault.setRelayerApproval.selector, msg.sender, relayer, approved),
            authorisation
        );
```

Reply:
I think I clarified this for myself. Relayer is a proxy, basically outlined [here](https://docs.balancer.fi/sdk/technical-reference/relayer.html#:~:text=Relayers%20are%20(user%20opt%2Din%2C%20audited)%20contracts%20that%20can%20make%20calls%20to%20the%20vault%20(with%20the%20transaction%20%E2%80%9Csender%E2%80%9D%20being%20any%20arbitrary%20address)%20and%20use%20the%20sender%E2%80%99s%20ERC20%20vault%20allowance%2C%20internal%20balance%20or%20BPTs%20on%20their%20behalf.). So giving the `vault` the approval and revoking it at the end of the tx is right.